### PR TITLE
Add a rank column

### PR DESCRIPTION
**Pull request by rainwoodman**
this is useful for abundance matching.

### CODE REVIEW

File: nbodykit/base/catalog.py [hunk 1]
```diff
@@ -1022,6 +1022,15 @@ def Weight(self):
         return ConstantArray(1.0, self.size, chunks=_global_options['dask_chunk_size'])
 
     @column
+    def Rank(self):
+        """
+        The column giving the rank of each particle in the list.
+
+        """
+        offset = sum(self.comm.allgather(self.size)[:self.comm.rank])
+        return da.arange(offset, offset + self.size, chunks=_global_options['dask_chunk_size'])
```
Comment (nickhand) [inline nbodykit/base/catalog.py]: use 'u4' as the dtype?

**Comment by nickhand** [comment]
I think we also use the "Rank" column here in the CGM algorithm:

https://github.com/bccp/nbodykit/blob/master/nbodykit/algorithms/cgm.py#L158

**Comment by nickhand** [comment]
and we should update the docs table of default columns:

https://github.com/bccp/nbodykit/blob/master/docs/source/catalogs/overview.rst

File: nbodykit/base/catalog.py (hunk 1 above)
Comment (rainwoodman) [reply nbodykit/base/catalog.py]: u8? It can be more than 4billion..

**Comment by rainwoodman** [comment]
Sure. BTW an issue with these default columns is that there is no easy way to write a catalog to disk bypassing these columns if they haven't been overridden.

**Comment by nickhand** [comment]
That's a good point. I guess the most sensible behavior would be not save those columns automatically and restore them when loading automatically. We would need to tag them somehow as defaults I guess..

**Comment by rainwoodman** [comment]
We currently have this 'hardcolumn' thing. But I keep forgetting what it really means.

**Comment by nickhand** [comment]
I believe the “hardcolumn” refers to any column that uses a method of the
class to return the column, so these default columns are “hardcolumns” but
not all hardcolumns are defaults

**Comment by rainwoodman** [comment]
I have been a bit wary about this. Looks like there are two ways of dealing with hardcolumns, either via a decorator or via a two function protocol. Do you know which subclass uses the protocol method and why?

**Comment by nickhand** [comment]
what do you mean by two function protocol? I think only the decorator is being used?

**Comment by rainwoodman** [comment]
There is also get_hardcolumns() that states it can be overridden...

**Comment by nickhand** [comment]
It looks like only the array and file catalogs overload get_hardcolumn() to dynamically get any of the fields in `dtype` when the user requests... I feel like we could probably simplify this to behave like a dictionary with default values?

**Comment by rainwoodman** [comment]
It currently already behave like a dict with default values. I think what we need is a tag on a column that it shall not be saved. Alternatively speaking, we want a function that returns a list of columns that shall go to disk.

**Comment by rainwoodman** [comment]
The rank column is different from others. I think it shouldn't be a column at all! Because after selection the rank changes!

**Comment by nickhand** [comment]
that is an excellent point

**Comment by rainwoodman** [comment]
Merge this after it is green? Chirag wants to use it.

**Comment by nickhand** [comment]
sorry... yes looks good to me. I'll merge

**Pull request merged**